keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1187 @@
+allows the response payloads to be returned directly to the user, without
+persisting them in etcd.
+
+#### Securing The CallBack Listener
+
+The `VolumeSnapshotDelta` callback listener will be secured by delegating the
+request authentication and authorisation to the Kubernetes API server, using the
+[`TokenReview`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#tokenreview-v1-authentication-k8s-io)
+and
+[`SubjectAccessReview`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#subjectaccessreview-v1-authorization-k8s-io)
+APIs to protect both the `nonResourceURL` endpoint as well as the resources.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The proposed design involves extending CSI with the `VolumeSnapshotDelta` CRD.
+Storage providers can opt in to support this feature by implementing the
+`VOLUME_SNAPSHOT_DELTA_SERVICE` capability in their CSI drivers.
+
+The `VolumeSnapshotDelta` resource is a namespace-scoped resource. It must be
+created in the same namespace as the base and target CSI `VolumeSnapshot`s.
+
+### CBT Datapath Worklow
+
+A Kubernetes user (can be a human or a computer process) initiates the datapath
+workflow by creating a new `VolumeSnapshotDelta` custom resource. The
+`VolumeSnapshotDelta` controller, deployed as a CSI sidecar in the storage
+provider's CSI driver, watches for new `VolumeSnapshotDelta` resources:
+
+![CBT Step 1](./img/cbt-step-01.png)
+
+Instead of satisfying the request directly, the controller constructs a callback
+URL and adds it to the resource's `status` subresource. The user can then use
+the callback URL to fetch the list of changed block metadata:
+
+![CBT Step 2](./img/cbt-step-02.png)
+
+The callback URL points to the `VolumeSnapshotDelta` listener, which runs in
+the same container as the `VolumeSnapshotDelta` controller. The CSI driver
+`Service` resource's FQDN will be used as the hostname of the callback URL, which
+essentially looks like:
+
+```sh
+https://<csi-driver.ns.svc.cluster.local>:<listener-port>/<resource-namespace>/<resource-name>
+```
+
+The `VolumeSnapshotDelta` listener delegates the authentication and
+authorisation of the request to the Kubernetes API server via the `TokenReview`
+and `SubjectAccessReview` APIs, in order to protect both the `nonResourceURL`
+endpoint as well as access to the `VolumeSnapshotDelta` resources.
+
+The user must include the `Authorization` header in the request, using an
+authorised service account's secret token as the bearer token. If the user's
+service account is deployed with `automountServiceAccountToken` set to `false`,
+they will have to extract the appropriate token from their secret.
+
```

Comment:
Where the sidecar gets the cluster TLS CA certificate and key for the HTTPS server?

Reply:
BTW, I think plain HTTP could be OK during alpha, still, you should have an idea how to add HTTPS.

Reply:
Yeah, I left out the TLS part for alpha. As far as the sidecar is concerned, I am thinking it can get its CA cert, ready-to-use signed cert etc. from a `Secret`. Essentially, the sidecar doesn't have to worry about issuing CSR. The bigger question is how that secret gets created (e.g., cert-manager). Can we offer a default self-signed approach, and an advance "bring your own certs" approach?

Reply:
I am not sure how other Kubernetes vendors handle certificates, IMO there is no common patter. OpenShift will generate a secret with cert + key for a Service, if the Service has a specific annotation.